p2p/src/net/mock/transport/impls/stream_adapter/wrapped_transport/utils.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::net::mock::transport::{impls::stream_adapter::traits::StreamAdapter, TransportSocket};
+
+use std::{pin::Pin, task::Poll};
+
+use futures::{future::BoxFuture, Future};
+
+use crate::Result;
+
+// Helper future used to drive handshakes concurrently.
+// It works like FuturesUnordered but instead returning Poll::Ready(None) if empty it will return Poll::Pending.
+pub struct HandshakeFut<'a, S: StreamAdapter<T::Stream>, T: TransportSocket>(
+    #[allow(clippy::type_complexity)]
+    pub  &'a mut Vec<(BoxFuture<'static, Result<S::Stream>>, T::Address)>,
+);
+
+impl<'a, S: StreamAdapter<T::Stream>, T: TransportSocket> Future for HandshakeFut<'a, S, T> {
+    type Output = (S::Stream, T::Address);
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+        'outer: loop {
+            for i in 0..self.0.len() {
+                match Future::poll(self.0[i].0.as_mut(), cx) {
+                    Poll::Ready(res) => {
+                        let (_, addr) = self.0.remove(i);
```

Comment:
This seems dangerous for a few reasons:
1. It seems that the Vec growth is unbounded.
2. Couple the unbounded growth with that we're looping over the whole vector in every poll, making this unbounded in execution time, opening nodes to remote DDoS.
3. Finally, to make all this worse, we remove arbitrary elements from the middle of a Vec, causing huge clones all the time...

The algorithm is OK, but the data structuring used is very bad. I can suggest a simple solution to removal, which is swapping element `i` and the last element of the Vec, but then I can imagine pinning might be an issue, besides the complexity in doing this.

We have to think of a better to do all this. Vec is wrong here.

- Worst case scenario, we should make the size of this Vec bounded.
- Best case sceanrio, we should be able to pop handshakes in O(1) time.

Reply:
The number of concurrent handshakes is limited to 100 and `Vec::swap_remove` is used instead (which runs in O(1)).
So all points should be addresses now.